autogpt_platform/backend/backend/notifications/notifications.py

```diff
@@ -0,0 +1,214 @@
+        """Queue a notification - exposed method for other services to call"""
+        try:
+            logger.info(f"Recieved Request to queue {event=}")
+            # Workaround for not being able to seralize generics over the expose bus
+            parsed_event = NotificationEventModel[
+                get_data_type(event.type)
+            ].model_validate(event.model_dump())
+            routing_key = self.get_routing_key(parsed_event)
+            message = parsed_event.model_dump_json()
+
+            logger.info(f"Recieved Request to queue {message=}")
+
+            exchange = "notifications"
+
+            # Publish to RabbitMQ
+            self.run_and_wait(
+                self.rabbit.publish_message(
+                    routing_key=routing_key,
+                    message=message,
+                    exchange=next(
+                        ex for ex in self.rabbit_config.exchanges if ex.name == exchange
+                    ),
+                )
+            )
+
+            return NotificationResult(
+                success=True,
+                message=(f"Notification queued with routing key: {routing_key}"),
+            )
+
+        except Exception as e:
+            logger.error(f"Error queueing notification: {e}")
+            return NotificationResult(success=False, message=str(e))
+
+    async def _process_immediate(self, message: str) -> bool:
+        """Process a single notification immediately"""
+        try:
+            event = NotificationEventDTO.model_validate_json(message)
+            parsed_event = NotificationEventModel[
+                get_data_type(event.type)
+            ].model_validate_json(message)
+            # Implementation of actual notification sending would go here
+            # self.email_sender.send_templated(event.type, event.user_id, parsed_event)
+            logger.info(f"Processing notification: {parsed_event}")
+            return True
+        except Exception as e:
+            logger.error(f"Error processing notification: {e}")
+            return False
+
+    def run_service(self):
+        logger.info(f"[{self.service_name}] Started notification service")
+
+        # Set up queue consumers
+        channel = self.run_and_wait(self.rabbit.get_channel())
+
+        immediate_queue = self.run_and_wait(
+            channel.get_queue("immediate_notifications")
+        )
+
+        while self.running:
+            try:
+                # Process immediate notifications
+                try:
+                    message = self.run_and_wait(immediate_queue.get())
+
+                    if message:
+                        success = self.run_and_wait(
+                            self._process_immediate(message.body.decode())
+                        )
+                        if success:
+                            self.run_and_wait(message.ack())
+                        else:
+                            self.run_and_wait(message.reject(requeue=True))
```

Comment:
Could this not cause an infinite loop, of requeueing failed messages, which are then reprocessed, which fails and requeue's them again. 

A mechanism to avoid this inf loop needs to be added.

Reply:
The dead letter + failed queue mechanism above handles that :)

Reply:
Failed messages get put into the failed queue to be reprocessed later (future PR)

Reply:
Same comment about tracking the need for change in a future PR